builtin/credential/aws/path_role.go

```diff
@@ -241,6 +282,16 @@ func (b *backend) nonLockedAWSRole(s logical.Storage, roleName string) (*awsRole
 		}
 	}
 
+	// Check if there was no pre-existing AuthType set (from older versions)
```

Comment:
Actually, I think it's a bit more complicated than that. The reason is that between releasing the read lock and acquiring the write lock, another thread might have acquired the write lock and updated the role entry out from under us. So we could be overwriting with stale data. Anyway, I just pushed a fix that saves the role entry and should be safe.